### PR TITLE
Discovery: Added frameworkSpec and different types used for discovering frameworks

### PR DESCRIPTION
**Pull request by svnsairam**
### Description

Creating small PR's for review. Removed some frameworksSpec and will be added in later PR's.
1. frameworkSpecs.ts : contains the specs for different frameworks.
2. types.ts : defines different interfaces used by the framework discovery.

### CODE REVIEW

File: src/frameworks/compose/discover/types.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  // Commands to run right before exporting the container image
+  // e.g. npm prune --omit=dev, yarn install --production=true
+  exportCommands?: string[];
+
+  // The runtime has detected a command that should always be run irrespective of
+  // the framework (e.g. the "build" script always wins in Node)
+  detectedCommands?: LifecycleCommands;
+
+  fallbackCommands?: LifecycleCommands;
```
Comment (taeold) [inline src/frameworks/compose/discover/types.ts:88]: can you explain what this command is meant to be?

File: src/frameworks/compose/discover/types.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  name: string;
+  description: string;
+  icon: string;
+
+  // oneof
+  install: SampleInstallStrategy;
+
+  // e.g. "javascript" and "typescript" variants
+  installVariants: Record<string, SampleInstallStrategy>;
+}
```
Comment (taeold) [inline src/frameworks/compose/discover/types.ts:106]: let's get rid of this if this isn't needed for our first cut.

File: src/frameworks/compose/discover/types.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  match(fs: FileSystem): Promise<boolean | null>;
+  getRuntimeName(): string;
+  analyseCodebase(fs: FileSystem, allFrameworkSpecs: FrameworkSpec[]): Promise<RuntimeSpec | null>;
+}
+
+export interface Command {
+  // Consider: string[] for series of commands that must execute successfully
+  // in sequence.
+  cmd: string | string[];
+  script?: string;
```
[the remark this answers is not in the document]
Comment (svnsairam) [reply src/frameworks/compose/discover/types.ts:16]: I will remove of these. I just kept these as optional as we are not using them any time soon.

File: src/frameworks/compose/discover/types.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+export interface Runtime {
+  match(fs: FileSystem): Promise<boolean | null>;
+  getRuntimeName(): string;
+  analyseCodebase(fs: FileSystem, allFrameworkSpecs: FrameworkSpec[]): Promise<RuntimeSpec | null>;
+}
+
+export interface Command {
+  // Consider: string[] for series of commands that must execute successfully
```
[the remark this answers is not in the document]
Comment (svnsairam) [reply src/frameworks/compose/discover/types.ts:13]: Do you mean script here?

File: src/frameworks/compose/discover/types.ts (hunk 1 above)
Comment (svnsairam) [reply src/frameworks/compose/discover/types.ts]: Removing fallbackCommands for now.

File: src/frameworks/compose/discover/types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (taeold) [reply src/frameworks/compose/discover/types.ts:13]: woops, i meant `string[]`

**Review by taeold — APPROVED** [review]

**Review by taeold — APPROVED** [review]

**Comment by svnsairam** [comment]
Merge PR on Discovery: Added frameworkSpec and different types used for discovering frameworks.

**Pull request closed without merging**